Don't choke on `map`, `splice`, and other array methods + make it reactive

DmitrySharabin opened the pull request.

DmitrySharabin committed:
Make it work

LeaVerou reviewed (changes requested):
This is a somewhat awkward solution. Instead of brute forcing it with forceUpdate and having to pass awkward options that don}t mean anything, I would prefer to first *understand* WHY it’s not updating. I suspect there is a better solution.

DmitrySharabin commented on js/RepoLabels.js:
```diff
@@ -5,14 +5,18 @@ let backends = {};
-			}
-		});
+
+		if (options) {
```
We need this to ensure we create an actual backends and do not perform any array operation, such as `map`, `splice`, etc., while working with an instance of `RepoLabels` class.

LeaVerou replied:
Do not add GitHub comments that explain what code does — that’s what code comments are for!

DmitrySharabin commented:
> I would prefer to first _understand_ WHY it’s not updating.

Probably a bug in Vue? https://codepen.io/dmitrysharabin/pen/ZEdQdpX

DmitrySharabin commented:
> I suspect there is a better solution.

So do I. Still can't find it, though 🤷‍♂️

DmitrySharabin committed:
Add comment

DmitrySharabin committed:
Get rid of the `context` option

DmitrySharabin committed:
Restore the previous code to leave only needed changes in this PR

DmitrySharabin commented:
I refactored this PR so the other parts would work correctly when the view update issue is resolved.

LeaVerou commented:
> > I would prefer to first _understand_ WHY it’s not updating.
> 
> Probably a bug in Vue? [codepen.io/dmitrysharabin/pen/ZEdQdpX](https://codepen.io/dmitrysharabin/pen/ZEdQdpX)

Have you reported it?

DmitrySharabin commented:
> Have you reported it?

No. I needed you to take a look at this first. It might be I don't see something.

LeaVerou commented:
> > Have you reported it?
> 
> No. I needed you to take a look at this first. It might be I don't see something.

Nothing jumps at me

DmitrySharabin committed:
Move side effects to a separate method

Inside `constructor()`, `this` is bound to the original object, not the reactive Proxy object. All side effects should be moved outside `constructor()` to make the Vue reactivity system work.

See https://stackoverflow.com/a/67895251/5990035

DmitrySharabin committed:
Fix buttons' hints

DmitrySharabin commented:
> > > Have you reported it?
> > 
> > 
> > No. I needed you to take a look at this first. It might be I don't see something.
> 
> Nothing jumps at me

I researched a bit more, and it turned out this is not a bug. It's an expected behavior. Inside `constructor()`, `this` is bound to the original object, not the reactive Proxy object. All side effects should be moved outside `constructor()` to make the Vue reactivity system work.

Some details are here: https://stackoverflow.com/a/67895251/5990035

So, here is another stab at fixing the app.

DmitrySharabin committed:
Add comment

DmitrySharabin committed:
Add semicolon

LeaVerou commented on js/index.js:
```diff
@@ -20,6 +20,7 @@ globalThis.app = createApp({
 		for (let name of this.repoNames) {
 			// this.load(name);
 			this.labels[name] = new RepoLabels(name, {syncWith: this.firstBackend});
+			this.labels[name].init();
```
I knew you could do it!
I’m assuming if we call `this.init()` from within the constructor it will have the same issue?

DmitrySharabin replied:
> I knew you could do it!

Thank you! 😊

> I’m assuming if we call `this.init()` from within the constructor it will have the same issue?

Yup. I also checked it to make sure it won't work. :)

LeaVerou commented on js/RepoLabels.js:
```diff
@@ -5,13 +5,19 @@ let backends = {};
+	init () {
+		// Move all side effects outside “constructor()” to make the Vue reactivity system work:
+		// “this” should be bound to the reactive Proxy object, not the original one
+		this.backend.load().then(labels => {
```
We can probably make this function async since it's a separate function. I had written it like that because constructors can't be async

DmitrySharabin committed:
Refactor to `async`/`await`

DmitrySharabin committed:
This shouldn't be async

We don't need to await init() to resolve—Vue will do everything for us.

Moreover, if mounted() is async, it breaks the `<madata-auth>` component.
